pyop2/compilation.py

```diff
@@ -112,26 +138,38 @@ def get_so(self, src, extension):
         # atomically (avoiding races).
         tmpname = os.path.join(cachedir, "%s_p%d.so.tmp" % (basename, pid))
 
+        if MPI.version >= 3:
+            # Assume that the library supports flashy MPI3 features
+            newcomm = self.comm.Split_type(MPI.COMM_TYPE_SHARED)
+            debug("Using MPI3 on rank ", newcomm.rank)
+            if newcomm.rank == 0:
+                singleton = True
+        else:
+            newcomm = self.comm
+            debug("Not using MPI3 on %d" % newcomm.rank)
+            singleton = self.single_out_a_rank(newcomm)
+
         if configuration['check_src_hashes'] or configuration['debug']:
-            matching = self.comm.allreduce(basename, op=_check_op)
+            matching = newcomm.allreduce(basename, op=_check_op)
```

Comment:
In the non-MPI three case, these barriers are all bigger than they need to be (because the comm is COMM_WORLD).  Hence my suggestion to split in both cases.